File: educational-core/src/com/jetbrains/edu/coursecreator/settings/CCOptions.java
package com.jetbrains.edu.coursecreator.settings;

import com.intellij.openapi.application.Experiments;
import com.intellij.openapi.ui.LabeledComponent;
import com.intellij.openapi.wm.IdeFocusManager;
import com.intellij.ui.components.JBCheckBox;
import com.jetbrains.edu.coursecreator.actions.CCPluginToggleAction;
import com.jetbrains.edu.learning.EduExperimentalFeatures;
import com.jetbrains.edu.learning.settings.OptionsProvider;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;

public class CCOptions implements OptionsProvider {
  private JRadioButton myHtmlRadioButton;
  private JRadioButton myMarkdownRadioButton;
  private JPanel myPanel;
  private JPanel myCustomOptions;

  private JBCheckBox myShowSplitEditorCheckBox = new JBCheckBox(null, CCSettings.getInstance().showSplitEditor());

  @Nullable
  @Override
  public JComponent createComponent() {
    if (!CCPluginToggleAction.isCourseCreatorFeaturesEnabled()) return null;
    if (CCSettings.getInstance().useHtmlAsDefaultTaskFormat()) {
      myHtmlRadioButton.setSelected(true);
      IdeFocusManager.getGlobalInstance().doWhenFocusSettlesDown(
        () -> IdeFocusManager.getGlobalInstance().requestFocus(myHtmlRadioButton, true));
    }
    else {
      myMarkdownRadioButton.setSelected(true);
      IdeFocusManager.getGlobalInstance().doWhenFocusSettlesDown(
        () -> IdeFocusManager.getGlobalInstance().requestFocus(myMarkdownRadioButton, true));
    }
    if (Experiments.isFeatureEnabled(EduExperimentalFeatures.SPLIT_EDITOR)) {
      LabeledComponent<JBCheckBox> showSplitEditorComponent =
        LabeledComponent.create(myShowSplitEditorCheckBox, "Show previous task file in framework lessons", BorderLayout.WEST);
      myCustomOptions.add(showSplitEditorComponent);
    }
    return myPanel;
  }

  @Override
  public boolean isModified() {
    CCSettings settings = CCSettings.getInstance();
    return myHtmlRadioButton.isSelected() != settings.useHtmlAsDefaultTaskFormat() ||
           myShowSplitEditorCheckBox.isSelected() != settings.showSplitEditor();
  }

  @Override
  public void apply() {
    if (isModified()) {
      CCSettings.getInstance().setUseHtmlAsDefaultTaskFormat(myHtmlRadioButton.isSelected());
      CCSettings.getInstance().setShowSplitEditor(myShowSplitEditorCheckBox.isSelected());
    }
  }

  @Override
  public void reset() {
    CCSettings settings = CCSettings.getInstance();
    myHtmlRadioButton.setSelected(settings.useHtmlAsDefaultTaskFormat());
    myMarkdownRadioButton.setSelected(!settings.useHtmlAsDefaultTaskFormat());
    myShowSplitEditorCheckBox.setSelected(settings.showSplitEditor());
  }

  @Override
  public void disposeUIResources() {

  }

  @Nls
  @Override
  public String getDisplayName() {
    return "Course Creator options";
  }
}

File: educational-core/src/com/jetbrains/edu/coursecreator/settings/CCOptions.kt
package com.jetbrains.edu.coursecreator.settings

import com.intellij.openapi.util.NlsSafe
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBRadioButton
import com.intellij.ui.dsl.builder.panel
import com.jetbrains.edu.coursecreator.actions.CCPluginToggleAction.Companion.isCourseCreatorFeaturesEnabled
import com.jetbrains.edu.learning.EduExperimentalFeatures
import com.jetbrains.edu.learning.isFeatureEnabled
import com.jetbrains.edu.learning.messages.EduCoreBundle
import com.jetbrains.edu.learning.settings.OptionsProvider
import javax.swing.ButtonGroup
import javax.swing.JComponent

@Suppress("UnstableApiUsage")
@NlsSafe
private const val HTML = "Html"

@Suppress("UnstableApiUsage")
@NlsSafe
private const val MARKDOWN = "Markdown"

class CCOptions : OptionsProvider {

  private val htmlRadioButton = JBRadioButton(HTML, CCSettings.getInstance().useHtmlAsDefaultTaskFormat())
  private val markdownRadioButton = JBRadioButton(MARKDOWN, !CCSettings.getInstance().useHtmlAsDefaultTaskFormat())

  private val copyTestsCheckBox = JBCheckBox(
    EduCoreBundle.message("ccoptions.copy.tests"),
    CCSettings.getInstance().copyTestsInFrameworkLessons()
  ).apply {
    toolTipText = EduCoreBundle.message("ccoptions.copy.tests.tooltip")
  }

  private val showSplitEditorCheckBox = JBCheckBox(
    EduCoreBundle.message("ccoptions.split.editor"),
    CCSettings.getInstance().showSplitEditor()
  )

  init {
    val buttonGroup = ButtonGroup()
    buttonGroup.add(htmlRadioButton)
    buttonGroup.add(markdownRadioButton)
  }

  override fun getDisplayName(): String = EduCoreBundle.message("ccoptions.display.name")

  override fun createComponent(): JComponent? {
    if (!isCourseCreatorFeaturesEnabled) return null

    return panel {
      group(displayName) {
        row {
          label(EduCoreBundle.message("ccoptions.description.format"))
          cell(htmlRadioButton)
          cell(markdownRadioButton)
        }
        row { cell(copyTestsCheckBox) }
        if (isFeatureEnabled(EduExperimentalFeatures.SPLIT_EDITOR)) {
          row { cell(showSplitEditorCheckBox) }
        }
      }
    }
  }

  override fun isModified(): Boolean {
    val settings = CCSettings.getInstance()
    return htmlRadioButton.isSelected != settings.useHtmlAsDefaultTaskFormat() ||
           showSplitEditorCheckBox.isSelected != settings.showSplitEditor() ||
           copyTestsCheckBox.isSelected != settings.copyTestsInFrameworkLessons()
  }

  override fun apply() {
    val settings = CCSettings.getInstance()
    settings.setUseHtmlAsDefaultTaskFormat(htmlRadioButton.isSelected)
    settings.setShowSplitEditor(showSplitEditorCheckBox.isSelected)
    settings.setCopyTestsInFrameworkLessons(copyTestsCheckBox.isSelected)
  }

  override fun reset() {
    val settings = CCSettings.getInstance()
    htmlRadioButton.isSelected = settings.useHtmlAsDefaultTaskFormat()
    markdownRadioButton.isSelected = !settings.useHtmlAsDefaultTaskFormat()
    showSplitEditorCheckBox.isSelected = settings.showSplitEditor()
    copyTestsCheckBox.isSelected = settings.copyTestsInFrameworkLessons()
  }
}
